File: push-be/internal/domain/usecase/calendar.go
package usecase

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"

	"push-be/internal/domain"
)

const maxCalendarRangeDays = 366

type CalendarService struct {
	events       domain.CalendarEventStore
	applications domain.ApplicationStore
}

func NewCalendarService(events domain.CalendarEventStore, applications domain.ApplicationStore) *CalendarService {
	return &CalendarService{events: events, applications: applications}
}

func validEventType(t domain.CalendarEventType) bool {
	switch t {
	case domain.CalInterview, domain.CalDeadline, domain.CalFollowUp, domain.CalCustom:
		return true
	}
	return false
}

func (s *CalendarService) List(ctx context.Context, userID uuid.UUID, f domain.CalendarEventFilter) (domain.Page[domain.CalendarEvent], error) {
	if f.From == nil || f.To == nil {
		return domain.Page[domain.CalendarEvent]{}, domain.Validation("from and to are required")
	}
	if !f.To.After(*f.From) {
		return domain.Page[domain.CalendarEvent]{}, domain.ValidationField("to", "must be after from")
	}
	if f.To.Sub(*f.From) > maxCalendarRangeDays*24*time.Hour {
		return domain.Page[domain.CalendarEvent]{}, domain.ValidationField("to", "range must be at most 366 days")
	}
	limit := f.Page.EffectiveLimit()
	items, err := s.events.ListCalendarEvents(ctx, userID, f, limit+1)
	if err != nil {
		return domain.Page[domain.CalendarEvent]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(e domain.CalendarEvent) domain.Cursor {
		return domain.Cursor{CreatedAt: e.CreatedAt, ID: e.ID}
	}), nil
}

type CreateEventInput struct {
	ApplicationID *uuid.UUID
	Type          domain.CalendarEventType
	Title         string
	StartsAt      time.Time
	EndsAt        time.Time
	TimeZone      string
	Notes         string
}

func (s *CalendarService) Create(ctx context.Context, userID uuid.UUID, in CreateEventInput) (*domain.CalendarEvent, error) {
	if err := validateEventInput(in.Type, in.Title, in.StartsAt, in.EndsAt, in.TimeZone); err != nil {
		return nil, err
	}
	if in.ApplicationID != nil {
		if _, err := s.applications.Get(ctx, userID, *in.ApplicationID); err != nil {
			if errors.Is(err, domain.ErrNotFound) {
				return nil, domain.NotFound()
			}
			return nil, domain.Internal()
		}
	}
	tz := in.TimeZone
	if tz == "" {
		tz = "UTC"
	}
	now := time.Now().UTC()
	e := &domain.CalendarEvent{
		ID: uuid.New(), UserID: userID, Revision: 1,
		ApplicationID: in.ApplicationID, Type: in.Type, Title: in.Title,
		StartsAt: in.StartsAt, EndsAt: in.EndsAt, TimeZone: tz,
		Source: domain.EventSourceLocal, Notes: in.Notes,
		CreatedAt: now, UpdatedAt: now,
	}
	if err := s.events.CreateCalendarEvent(ctx, e); err != nil {
		return nil, domain.Internal()
	}
	return e, nil
}

func validateEventInput(t domain.CalendarEventType, title string, startsAt, endsAt time.Time, tz string) *domain.Error {
	if !validEventType(t) {
		return domain.ValidationField("type", "unsupported event type")
	}
	if title == "" || len(title) > 300 {
		return domain.ValidationField("title", "title must be 1-300 characters")
	}
	if startsAt.IsZero() {
		return domain.ValidationField("startsAt", "required")
	}
	if !endsAt.After(startsAt) {
		return domain.ValidationField("endsAt", "must be after startsAt")
	}
	if tz != "" {
		if _, err := time.LoadLocation(tz); err != nil {
			return domain.ValidationField("timeZone", "invalid IANA time zone")
		}
	}
	return nil
}

type PatchEventInput struct {
	ExpectedRevision int64
	Title            *string
	StartsAt         *time.Time
	EndsAt           *time.Time
	TimeZone         *string
	Notes            *string
}

func (s *CalendarService) Patch(ctx context.Context, userID, id uuid.UUID, in PatchEventInput) (*domain.CalendarEvent, error) {
	e, err := s.events.GetCalendarEvent(ctx, userID, id)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil, domain.NotFound()
		}
		return nil, domain.Internal()
	}
	if e.Revision != in.ExpectedRevision {
		return nil, domain.RevisionConflict(e.Revision)
	}
	if in.Title != nil {
		if *in.Title == "" || len(*in.Title) > 300 {
			return nil, domain.ValidationField("title", "title must be 1-300 characters")
		}
		e.Title = *in.Title
	}
	if in.StartsAt != nil {
		e.StartsAt = *in.StartsAt
	}
	if in.EndsAt != nil {
		e.EndsAt = *in.EndsAt
	}
	if !e.EndsAt.After(e.StartsAt) {
		return nil, domain.ValidationField("endsAt", "must be after startsAt")
	}
	if in.TimeZone != nil {
		if _, err := time.LoadLocation(*in.TimeZone); err != nil {
			return nil, domain.ValidationField("timeZone", "invalid IANA time zone")
		}
		e.TimeZone = *in.TimeZone
	}
	if in.Notes != nil {
		e.Notes = *in.Notes
	}
	e.UpdatedAt = time.Now().UTC()
	if err := s.events.PatchCalendarEvent(ctx, e, in.ExpectedRevision); err != nil {
		return nil, mapRevisionErr(err)
	}
	e.Revision = in.ExpectedRevision + 1
	return e, nil
}

func (s *CalendarService) Delete(ctx context.Context, userID, id uuid.UUID, expectedRevision int64) error {
	if err := s.events.DeleteCalendarEvent(ctx, userID, id, expectedRevision); err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return domain.NotFound()
		}
		return mapRevisionErr(err)
	}
	return nil
}

File: push-be/app/domain/services/calendar.py
from __future__ import annotations
from datetime import datetime, timedelta, timezone
from typing import Optional
from uuid import UUID, uuid4
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from app.db import DB, NotFoundError
from app.domain import entities as ent
from app.domain.errors import (
    internal, map_revision_err, not_found, revision_conflict, validation,
    validation_field,
)
from app.infrastructure.store_applications import ApplicationStore
from app.infrastructure.store_calendar import CalendarEventStore

MAX_RANGE_DAYS = 366


def _now() -> datetime:
    return datetime.now(timezone.utc)


def _valid_tz(tz: str) -> bool:
    try:
        ZoneInfo(tz)
        return True
    except (ZoneInfoNotFoundError, ValueError, KeyError):
        return False


def validate_event_input(type_: str, title: str, starts_at: datetime,
                         ends_at: datetime, tz: str) -> None:
    if not ent.valid_event_type(type_):
        raise validation_field("type", "unsupported event type")
    if not 1 <= len(title) <= 300:
        raise validation_field("title", "title must be 1-300 characters")
    if starts_at is None:
        raise validation_field("startsAt", "required")
    if ends_at <= starts_at:
        raise validation_field("endsAt", "must be after startsAt")
    if tz and not _valid_tz(tz):
        raise validation_field("timeZone", "invalid IANA time zone")


class CalendarService:
    def __init__(self, db: DB):
        self.db = db
        self.events = CalendarEventStore(db)
        self.applications = ApplicationStore(db)

    async def list(self, user_id: UUID, application_id: Optional[UUID],
                   from_: Optional[datetime], to: Optional[datetime],
                   source: Optional[str], page):
        if from_ is None or to is None:
            raise validation("from and to are required")
        if to <= from_:
            raise validation_field("to", "must be after from")
        if to - from_ > timedelta(days=MAX_RANGE_DAYS):
            raise validation_field("to", "range must be at most 366 days")
        try:
            return await self.events.list(user_id, application_id, from_, to,
                                          source, page)
        except Exception:
            raise internal()

    async def create(self, user_id: UUID, application_id: Optional[UUID],
                     type_: str, title: str, starts_at: datetime, ends_at: datetime,
                     time_zone: str, notes: str) -> ent.CalendarEvent:
        validate_event_input(type_, title, starts_at, ends_at, time_zone)
        if application_id is not None:
            try:
                await self.applications.get(user_id, application_id)
            except NotFoundError:
                raise not_found()
            except Exception:
                raise internal()
        now = _now()
        e = ent.CalendarEvent(
            id=uuid4(), user_id=user_id, revision=1, application_id=application_id,
            type=type_, title=title, starts_at=starts_at, ends_at=ends_at,
            time_zone=time_zone or "UTC", source=ent.EVENT_SOURCE_LOCAL,
            notes=notes, created_at=now, updated_at=now)
        try:
            await self.events.create(e)
        except Exception:
            raise internal()
        return e

    async def patch(self, user_id: UUID, id_: UUID, expected: int,
                    title: Optional[str], starts_at, ends_at,
                    time_zone: Optional[str], notes: Optional[str]) -> ent.CalendarEvent:
        try:
            e = await self.events.get(user_id, id_)
        except NotFoundError:
            raise not_found()
        except Exception:
            raise internal()
        if e.revision != expected:
            raise revision_conflict(e.revision)
        if title is not None:
            if not 1 <= len(title) <= 300:
                raise validation_field("title", "title must be 1-300 characters")
            e.title = title
        if starts_at is not None:
            e.starts_at = starts_at
        if ends_at is not None:
            e.ends_at = ends_at
        if e.ends_at <= e.starts_at:
            raise validation_field("endsAt", "must be after startsAt")
        if time_zone is not None:
            if not _valid_tz(time_zone):
                raise validation_field("timeZone", "invalid IANA time zone")
            e.time_zone = time_zone
        if notes is not None:
            e.notes = notes
        e.updated_at = _now()
        try:
            await self.events.patch(e, expected)
        except Exception as err:
            raise map_revision_err(err)
        e.revision = expected + 1
        return e

    async def delete(self, user_id: UUID, id_: UUID, expected: int) -> None:
        try:
            await self.events.delete(user_id, id_, expected)
        except NotFoundError:
            raise not_found()
        except Exception as err:
            raise map_revision_err(err)
